SOLR-16976: Remove log4j-jul jar and use slf4j bridge for JUL

szhou1998 opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16976

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

https://github.com/apache/solr/pull/1765 introduced a change that causes an exception when remote JMX is enabled (stack trace included in Jira description).

# Solution

@elyograg provided a patch which resolves this issue by removing log4j-jul and using the slf4j bridge for JUL. I removed the NOCOMMIT from the patch.

# Tests

@elyograg tested the patch. Added a test that there's no exception on Solr startup with remote JMX enabled.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended, not available for branches on forks living under an organisation)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

epugh commented:
I don't have much knowledge in how logging works.  One thing that would help, would you mind creating a BATS style test that demonstrates the bug and the fix?  I believe you could write a test that starts up solr with remote JMX, and then we see the bug, and then the fix.   The bats tests are in `./solr/packaging/test`.    Let me know if you have any questions!

szhou1998 commented:
> I don't have much knowledge in how logging works. One thing that would help, would you mind creating a BATS style test that demonstrates the bug and the fix? I believe you could write a test that starts up solr with remote JMX, and then we see the bug, and then the fix. The bats tests are in `./solr/packaging/test`. Let me know if you have any questions!

I've removed the patch and added a test to start solr with remote JMX enabled. It should fail with the exception seen in the Jira.

epugh commented:
Humm...  I ran the "Solr Script Tests" expecting it to fail...    I can run it locally tomorrow and see what I see!

szhou1998 commented:
Ah my mistake, Solr starts even though the exception is logged. Added a refute_output for the exception.

epugh commented:
okay, the refute wasn't actually matching..   i found that we need to consult the log file.  Also, I was getting a RMI Port exception..    Can you validate that this test fails for you, and hten add in your patch?

cpoerschke commented on solr/bin/solr:
```diff
@@ -1076,10 +1076,8 @@ if [ "${SOLR_HOME:0:${#EXAMPLE_DIR}}" = "$EXAMPLE_DIR" ]; then
 if [ -n "${LOG4J_PROPS:-}" ]; then
-  LOG4J_CONFIG+=("-Dlog4j.configurationFile=$LOG4J_PROPS")
+  LOG4J_CONFIG=("-Dlog4j.configurationFile=$LOG4J_PROPS")
 fi
```
Wondering if `else` is needed to ensure `LOG4J_CONFIG` is always initialised?
```suggestion
else
  LOG4J_CONFIG=
fi
```

epugh replied:
Do we have a bats test that would catch the `LOG4J_CONFIG` not being initialized and show if this is needed?  

There is a lot of tricky logic in the start command, and I'd love to have more tests so that as we refactor/change it we have more confidence that the tests are meaningful!

szhou1998 replied:
I added a commit to initialize LOG4J_CONFIG=() which is what it was originally https://github.com/apache/solr/pull/1765/files#diff-49c29c8f653341f48008c38f5f2cf970fa430cdca29181c819d7bdcfcc722980

epugh commented:
Okay, so, after putting in the slf4j bridge, I see that the bats tests is now failing (and thats a good thing) because it means we don't see the error:

```
not ok 85 SOLR-16976 solr starts with remote JMX enabled # in 7641 ms
# (from function `assert_output' in file /home/runner/work/solr/solr/.gradle/node/packaging/node_modules/bats-assert/src/assert.bash, line 247,
#  in test file test/test_start_solr.bats, line 40)
#   `assert_output --partial 'java.lang.ClassNotFoundException: org.apache.logging.log4j.jul.LogManager'' failed
```

@szhou1998 would you mind flipping the `assert_output` to a `refute_output` and lets see if the test now passes!

dsmiley commented on solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:
```diff
@@ -133,6 +134,9 @@ public SolrDispatchFilter() {}
   @Override
   public void init(FilterConfig config) throws ServletException {
     try {
+      SLF4JBridgeHandler.removeHandlersForRootLogger();
+      SLF4JBridgeHandler.install();
```
This logic should ideally be done as early as possible.  A ServletFilter's init method is not actually so early as we can do.  We could do CoreContainerProvider.init() which will start up CoreContainer _before_ the ServletFilter.  But we could _do earlier than that_ by **adding some XML based configuration in jetty.xml, which I think makes more sense**.  It could also mean we add the dependency JAR without formerly making it a dependency of solr-core, thus wouldn't leak to other/external things projects that might want to depend on solr-core.  This is a deployment matter so it's best done at the Jetty layer.

szhou1998 replied:
Would that change be in this file? https://github.com/apache/solr/blob/457ca3668a8906277f8f638af9c740b57f2a2a20/solr/server/etc/jetty.xml#L8
Do you know what it would look like?

dsmiley replied:
A quick google search turned up [this](https://github.com/rapla/rapla/blob/master/src/test/etc/jetty.xml) immediately.

dsmiley commented:
Please stop force-pushing to PRs -- it resets the review state in GitHub.
Any way, I will merge this soon; thank you!
